libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -786,42 +790,79 @@ void _serializeConnect( const IotMqttConnectInfo_t * pConnectInfo,
     }
 
     /* If metrics are enabled, write the metrics username into the CONNECT packet.
-     * Otherwise, write the username and password only when not connecting to an
+     * Otherwise, write the username and password only when not connecting to the
      * AWS IoT MQTT server. */
     if( pConnectInfo->awsIotMqttMode == true )
     {
-        #if AWS_IOT_MQTT_ENABLE_METRICS == 1
+        #if ( AWS_IOT_MQTT_ENABLE_METRICS == 1 ) && ( IOT_STATIC_MEMORY_ONLY == 0 )
             IotLogInfo( "Anonymous metrics (SDK language, SDK version) will be provided to AWS IoT. "
                         "Recompile with AWS_IOT_MQTT_ENABLE_METRICS set to 0 to disable." );
 
-            pBuffer = _encodeString( pBuffer,
-                                     AWS_IOT_METRICS_USERNAME,
-                                     AWS_IOT_METRICS_USERNAME_LENGTH );
-        #endif
+            /* Determine if the Connect packet should use a combination of the username
+             * for authentication plus the SDK version string. */
+            if( pConnectInfo->pUserName != NULL )
+            {
+                pcUserName = Iot_DefaultMalloc( pConnectInfo->userNameLength +
```

Comment:
Another buffer shouldn't be allocated for this.  The username can be encoded directly into `pBuffer`; you just can't use the `_encodeString` function as currently implemented to do this.  Or you could modify that function for this use case.